src/classes/CON_PreferredPhone_TDTM.cls

```diff
@@ -0,0 +1,77 @@
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+ 
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Contacts
+* @group-content ../../ApexDocContent/Contacts.htm
+* @description Populates default phone field according to user preferences and/or enforces use of 'Preferred Phone'.
+*/
+public class CON_PreferredPhone_TDTM extends TDTM_Runnable{
+/*******************************************************************************************************
+    * @description Populates default phone field according to user preferences and/or enforces use of 'Preferred Phone'
+    * @return dmlWrapper
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
+        TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+             
+        if (!TDTM_ProcessControl.getRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_PreferredPhone_TDTM) 
+            && (triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate) && newlist != null) {
+
+            // While this logic is not performing any DML currently, the recursion flag is utilized here to be ready for any future logic changes.
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.registeredTrigger.CON_PreferredPhone_TDTM, true);
+            
+            Boolean validatePreferredPhone = !UTIL_CustomSettingsFacade.getSettings().Disable_Preferred_Phone_Enforcement__c;
+            String selectedPreferredPhone = UTIL_CustomSettingsFacade.getSettings().Preferred_Phone_Selection__c;   
+ 
+            for (SObject so : newlist) {
+                Contact contact = (Contact)so;
+
+                // Process the Prefered Phone
+
+                if (oldList != NULL) {
+
+                     Map<ID, Contact> oldContactMap = new Map<ID, Contact>((List<Contact>)oldlist);
```

Comment:
We should store this map outside of the for loop so we don't need to rebuild it on every single iteration.

Then line 63 just becomes `CON_Phone.processPreferredPhone(contact, oldContactMap.get(contact.Id), validatePreferredPhone, selectedPreferredPhone);`

Reply:
Hmm if I put the map out side of FOR loop I still have to put int the  if (oldList != NULL) IF condition since it will break for inserts(old list would be NULL), also its scope is limited to the IF and won't be available in for loop.

Reply:
@skosgi1 But you wouldn't need to reference oldList anymore?

Try something like...

`Map<Id, Contact> oldContactMap;
if (oldList != null) {
    oldContactMap = new Map<ID, Contact>((List<Contact>)oldlist);
}

for (SObject so : newList) {
    Contact contact = (Contact)so;
    CON_Phone.processPreferredPhone(contact, oldContactMap.get(contact.Id), validatePreferredPhone, selectedPreferredPhone);
}

Reply:
(Note I edited this a few times - final version above)